docs/design_proposals/sync-auto-support.md

```diff
@@ -0,0 +1,212 @@
+COPY build/classes/java/main/ /app/classes
+
+CMD ["java", "-cp", "/app/classes:/app/dependencies/*", "hello.Application"]
+```
+
+with a skaffold sync block that looks like:
+```
+sync:
+  manual:
+    - src: "build/classes/java/main/**/*.class"
+      dest: "/app/classes"
+      strip: "build/classes/java/main/"
+```
+
+A user's devloop then looks like this:
+
+1. run `./gradlew classes copyDependencies`
+2. run `skaffold dev`
+3. *make changes to some java file*
+4. run `./gradlew classes`
+5. *skaffold syncs files*
+
+which is far from ideal.
+
+#### Fix it?
+
+What we might want, is an API that can accept configuration that looks like:
+
+##### User Configuration
+
+User's that use `auto` should expect the builder-sync to do the right thing and
+will not be required to do much configuration.
+
+Auto will only work with builders that have implemented the auto spec. We
+expect at least `jib` to do this.
+
+```yaml
+build:
+  artifacts:
+  - image: ...
+    context: jib-project
+    jib: {}
+    sync:
+      auto: {}
+```
+
+##### Sync Api with Buidler
+
+The builder will talk to sync component by providing it with the following data
+1. A list of `generated` configs, each containing
+    1. A `command` to generate files to sync
+    2. A list of inputs to watch as triggers
+    3. A list of syncs (src, dest) to execute after generation
+2. A list of `direct` sync directives (src, dest) to execute without any script
+   execution
+
+So maybe some datastructures like (I don't really know a lot of go, so assume
+this will be written in some consistent way eventually):
+
+```
+type AutoSync struct {
+  generated []Generated
+  direct []Direct
+}
+
+type Generated struct {
+  command []String
+  inputs []File
+  syncables []Syncables
+}
+
+type Direct struct {
+  syncables []Syncables
+}
+
+type Syncables struct {
+  src String
+  dest String
+  strip String
+}
+```
+
+###### Jib - Skaffold Sync interface
+
+How a tool like Jib might surface the necessary information to Skaffold
+
+I would expect to add a task like `_jibSkaffoldSyncMap` that will produce
+json output for the skaffold-jib intergration to consume and forward to the sync
+system. And example output might look like:
+
+```
+BEGIN JIB JSON
+{
+  "generated": {
+    “target/classes”: "app/classes",
+    "target": "dest2"
+  },
+  "direct": {
+    "src/main/extra1": "/",
+    "src/main/extra2": "/"
+    ".m2/some/dep/my-dep-SNAPSHOT.jar": "app/dependencies/my-dep-SNAPSHOT.jar"
+  }
+}
+```
+
+perhaps jib should also tell the system what task/goals are required to build
+the generated files??
+
+```
+{
+  ...
+  "generator-tasks": ["classes", "resources"]
+  "generated": ...
+}
+```
+
+
+### Open Issues/Questions
+
+Please list any open questions here in the following format:
+
+** What happens if the build definition changes? **
+I'm not sure if we should just allow the user to go nuts, or if we should
+explicitly disallow this, because changing the build definition might result in
```

Comment:
Current behavior is when a user changes `skaffold.yaml` when a `skaffold dev` loop is running, a build and deploy is triggered.
So, this should be fine?

Reply:
Yeah, I guess we can just pick a mode for now and define it, change it later if needed? If the java build definition changes, then that triggers a full rebuild too.